istresearch/traptor
# P4 1881 - Traptor Connection Exception Handling

# For the Reviewer

When this is complete, you should approve the PR via github.

# For the Reviewee

<Reminder PR Title should be JIRA_NUMBER and Useful Description>

## Summary
Re-implemented handling of Twitter connection exceptions. All BirdyExceptions are retried with a linear backoff starting at 60s, accumulating 60s each retry, for a max wait of 5 min to respect Twitter's back-off message and preserve our API quota. TwitterAuthError exceptions put the Traptor into an idle state and turn off it's heartbeat so that it's rules are reallocated to other functional Traptor instances. 

Other changes include signal handling and cleanup for all resources during shutdown. Additional Datadog metrics to improve monitoring of Traptor is also included.

#### Release Note
Re-implemented handling of Twitter connection exceptions.

#### Breaking Changes
<Description for Techops of how to handle changes, migrations, updates>None.

## Quality Assurance

You have gathered the following items:

Given all of the items above, you have updated your Application ORT at the following locations:
- **Features and Alerting**: <link to app ORT>Required.
- **P4 Alerting**: <link to p4-alerting PR>Required.

## Testing and Verification

1. Run stacks for Cooper and Traptor Rule Manager (docker-compose-2.yml): Cooper web and engine, Traptor RM, elasticsearch 7.8

2. Configure a `traptor.env` with valid Twitter credentials.

3. Run the Traptor dependencies: redis, kafka, twitterapi

4. Start traptor: `docker-compose up --build traptor`

5. Traptor should start up and begin waiting for rules. Kill Traptor `ctrl-c` to ensure it shuts down cleanly and quickly.

6. Start Traptor again. Activate a Cooper rule:
```
 $ curl -XPOST -H "content-type:application/json" "localhost:8080/cooper/rule" -d'{"type":"keyword", "value":"soccer"}'
```

7. Rule manager should process the rule and assign it to the Traptor. Traptor should receive a restart notification and initiate a Birdy connection to collect.

8. Restart Traptor many times. Let it run long enough to connect to Twitter, then immediately restart it again until you receive a `TwitterRateLimitError`. Once the error is recceived, leave Traptor running. Traptor should wait a full minute before retrying the connection. It should succeed.

9. Stop Traptor. Alter the API credentials to make them invalid. Start Traptor and wait for rule manager to assign the collection rule.

10. Traptor should receive a `TwitterAuthError` and enter an idle mode. The auth error should be logged once a minute. Also, the heartbeat should stop letting rule manager know to reassign the rules away. Watch the rule manager logs for this.
